nav2_map_server/src/map_3d/map_io_3d.cpp

```diff
@@ -0,0 +1,382 @@
+// Copyright (c) 2020 Shivam Pandey
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*Map-io library for 3D pointcloud implementation*/
+
+#include "nav2_map_server/map_3d/map_io_3d.hpp"
+
+#ifndef _WIN32
+#include <libgen.h>
+#endif
+#include <iostream>
+#include <string>
+#include <vector>
+#include <stdexcept>
+
+#include "yaml-cpp/yaml.h"
+#include "nav2_map_server/map_3d/pcl_helper.hpp"
+
+#include "pcl/io/pcd_io.h"
+#include "Eigen/Core"
+
+#ifdef _WIN32
+// https://github.com/rtv/Stage/blob/master/replace/dirname.c
+static
+char * dirname(char * path)
+{
+  static const char dot[] = ".";
+  char * last_slash;
+
+  if (path == NULL) {
+    return path;
+  }
+
+  /* Find last '/'.  */
+  last_slash = path != NULL ? strrchr(path, '/') : NULL;
+
+  if (last_slash != NULL && last_slash == path) {
+    /* The last slash is the first character in the string.  We have to
+       return "/".  */
+    ++last_slash;
+  } else if (last_slash != NULL && last_slash[1] == '\0') {
+    /* The '/' is the last character, we have to look further.  */
+    last_slash = reinterpret_cast<char *>(memchr(path, last_slash - path, '/'));
+  }
+
+  if (last_slash != NULL) {
+    /* Terminate the path.  */
+    last_slash[0] = '\0';
+  } else {
+    /* This assignment is ill-designed but the XPG specs require to
+       return a string containing "." in any case no directory part is
+       found and so a static and constant string is required.  */
+    path = reinterpret_cast<char *>(dot);
+  }
+
+  return path;
+}
+#endif
+
+namespace nav2_map_server
+{
+
+namespace map_3d
+{
+
```

Comment:
I largely skimmed this file. I expect that @AlexeyMerzlyakov has vetted this with proper error checking, etc